File: atlantic_report/atlantic_report/report/sales_report___detail/sales_report___detail.js
// Sales Report - Detail (Final Revised Patch)

frappe.query_reports["Sales Report - Detail"] = {

    onload(report) {

        const style = document.createElement("style");
        style.innerHTML = `

            /* =====================================================
               GLOBAL FONT 12PX
            ===================================================== */
            .dt-cell, .dt-cell__content, .dt-cell__text {
                font-size: 12px !important;
                line-height: 1.25 !important;
            }

            /* =====================================================
               HEADER ROW AUTO HEIGHT — ONLY HEADER
            ===================================================== */
            .dt-row-header {
                height: auto !important;
                min-height: auto !important;
                display: flex !important;
                align-items: flex-start !important;
            }

            .dt-row-header .dt-cell {
                height: auto !important;
                min-height: auto !important;
                display: flex !important;
                align-items: flex-start !important;
            }

            .dt-row-header .dt-cell__content {
                height: auto !important;
                min-height: auto !important;
                display: block !important;
                white-space: normal !important;
                line-height: 1.2 !important;
                overflow: visible !important;
            }

            /* =====================================================
               BODY ROW — DO NOT TOUCH HEADER OR FILTER
            ===================================================== */
            .dt-body .dt-row:not(.dt-row-header):not(.dt-row-filter) {
                height: 75px !important;
                min-height: 75px !important;
            }

            .dt-body 
            .dt-row:not(.dt-row-header):not(.dt-row-filter) 
            .dt-cell__wrapper,
            .dt-body 
            .dt-row:not(.dt-row-header):not(.dt-row-filter) 
            .dt-cell__content {
                height: 75px !important;
                min-height: 75px !important;
            }

            /* =====================================================
               FIX ITEM DISPLAY — ONLY BODY, NEVER HEADER/FILTER
            ===================================================== */
            .dt-row:not(.dt-row-header):not(.dt-row-filter) .dt-cell--item_display {
                overflow: visible !important;
            }

            .dt-row:not(.dt-row-header):not(.dt-row-filter) 
                .dt-cell--item_display .dt-cell__wrapper {
                height: auto !important;
                overflow: visible !important;
                display: block !important;
                align-items: flex-start !important;
            }

            .dt-row:not(.dt-row-header):not(.dt-row-filter) 
                .dt-cell--item_display .dt-cell__content {
                height: auto !important;
                overflow: visible !important;
                display: block !important;
                white-space: normal !important;
            }

            .dt-row:not(.dt-row-header):not(.dt-row-filter) 
                .dt-cell--item_display .dt-cell__text {
                white-space: normal !important;
                display: block !important;
                overflow: visible !important;
            }

            /* =====================================================
               RIGHT ALIGN — PRICE & AMOUNT ONLY IN BODY ROW
            ===================================================== */
            .dt-row:not(.dt-row-header):not(.dt-row-filter) 
                .dt-cell--rate .dt-cell__content,
            .dt-row:not(.dt-row-header):not(.dt-row-filter) 
                .dt-cell--rate .dt-cell__text,
            .dt-row:not(.dt-row-header):not(.dt-row-filter) 
                .dt-cell--amount .dt-cell__content,
            .dt-row:not(.dt-row-header):not(.dt-row-filter) 
                .dt-cell--amount .dt-cell__text {
                text-align: right !important;
                justify-content: flex-end !important;
                width: 100% !important;
                display: flex !important;
            }
        `;
        document.head.appendChild(style);

        /* =====================================================
           FORCE BODY ROW HEIGHT ONLY (HEADER/FILTER untouched)
        ===================================================== */
        /* =====================================================
        FORCE ALL ROW HEIGHT = 75px (HEADER, FILTER, BODY)
        ===================================================== */
        const patch_row_height = () => {
            const dt = frappe.query_report?.datatable;
            if (!dt) return;

            const H = 85;

            // Set option global
            dt.options.cellHeight = H;

            const rm = dt.rowmanager;
            if (!rm) return;

            // Simpan original jika suatu saat ingin dikembalikan
            if (!rm.__orig_getRowHeight) {
                rm.__orig_getRowHeight = rm.getRowHeight?.bind(rm);
            }

            // Override tinggi baris untuk SEMUA row
            rm.getRowHeight = function (rowIndex) {
                return H;
            };

            // Set properti internal lain
            rm.rowHeight = H;
            rm.row_height = H;

            // Refresh table
            try {
                rm.refreshRows(dt.datamanager.getRows());
            } catch (e) {
                dt.refresh();
            }
        };

        frappe.after_ajax(patch_row_height);
        setTimeout(patch_row_height, 400);
        setTimeout(patch_row_height, 1200);

    },

    filters: [
        {
            fieldname: "from_date",
            label: __("From Date"),
            fieldtype: "Date",
            default: frappe.datetime.month_start(),
            reqd: 1
        },
        {
            fieldname: "to_date",
            label: __("To Date"),
            fieldtype: "Date",
            default: frappe.datetime.get_today(),
            reqd: 1
        },
        {
            fieldname: "customer",
            label: __("Customer"),
            fieldtype: "Link",
            options: "Customer"
        }
    ],

    /* =====================================================
       FORMATTER (unchanged)
    ===================================================== */
    formatter(value, row, column, data, default_formatter) {

        /* ==========================================================
        KHUSUS KOL0M QTY_UOM (detail + total row)
        ========================================================== */
        if (column.fieldname === "qty_uom" && value) {
            var html = value || "";

            // Ambil angka qty
            var qty_match = html.match(/>([\d.,]+)</);
            var qty_val = qty_match ? qty_match[1] : "0";
            qty_val = format_number(flt(qty_val), null, 2);   // thousand separator + 2 decimal

            // Ambil UOM (detail saja)
            var uom_match = html.match(/<br>\s*<span[^>]*>(.*?)<\/span>/);
            var uom_val = uom_match ? uom_match[1] : "";

            // TOTAL ROW → hanya angka bold, no UOM
            if (data && data.is_total_row) {
                return `<div style="text-align:right;font-weight:bold;">${qty_val}</div>`;
            }

            // DETAIL ROW → Qty + UOM
            return `
                <div style="display:block;">
                    <div style="text-align:right;">${qty_val}</div>
                    ${uom_val ? `<div>${uom_val}</div>` : ``}
                </div>
            `;
        }

        /* ==========================================================
        BARIS TOTAL UNTUK KOLOM LAIN
        ========================================================== */
        if (data && data.is_total_row) {

            // -- Kolom PRICE harus dikosongkan --
            if (column.fieldname === "rate") {
                return "";
            }

            // -- Kolom AMOUNT → format currency 2 decimal --
            if (column.fieldname === "amount") {
                return `<b>${format_currency(flt(value), data.currency, 2)}</b>`;
            }

            // -- Kolom lainnya → bold value biasa --
            return `<b>${value || ""}</b>`;
        }

        /* ==========================================================
        CURRENCY (detail row)
        ========================================================== */
        if (column.fieldtype === "Currency") {
            return `
                <div style="text-align:right;width:100%;">
                    ${format_currency(flt(value), data.currency, 2)}
                </div>
            `;
        }

        /* ==========================================================
        FLOAT / INT (detail row)
        ========================================================== */
        if (["Float", "Int"].includes(column.fieldtype)) {
            return `
                <div style="text-align:right;width:100%;">
                    ${format_number(value, null, 2)}
                </div>
            `;
        }

        /* ==========================================================
        DEFAULT (text, HTML, dll.)
        ========================================================== */
        return default_formatter(value, row, column, data);
    }

};

File: atlantic_report/atlantic_report/report/sales_report___detail/sales_report___detail.py
# Copyright (c) 2025, Takwindo and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import formatdate, flt, get_url_to_form

def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    return columns, data

def get_columns():
    return [
        {
            "label": "DO No<br>Date<br>Salesman", 
            "fieldname": "do_info", 
            "fieldtype": "Data", 
            "width": 130,
            "align": "left"
        },
        {
            "label": "Invoice No<br>Date", 
            "fieldname": "inv_info", 
            "fieldtype": "Data", 
            "width": 110,
            "align": "left"
        },
        {
            "label": "Customer", 
            "fieldname": "customer_name", 
            "fieldtype": "Data", 
            "width": 230
        },
        {
            "label": "Item Code<br>Description (Brand, Model, Size)", 
            "fieldname": "item_display", 
            "fieldtype": "Data", 
            "width": 300
        },
        {
            "label": "Curr", 
            "fieldname": "currency", 
            "fieldtype": "Data", 
            "width": 50
        },
        {
            "label": "Qty<br>U/M", 
            "fieldname": "qty_uom", 
            "fieldtype": "Data", 
            "width": 80,
            "align": "right"
        },
        {
            "label": "Price", 
            "fieldname": "rate", 
            "fieldtype": "Currency", 
            "options": "currency",
            "width": 90
        },
        {
            "label": "Amount", 
            "fieldname": "amount", 
            "fieldtype": "Currency", 
            "options": "currency", 
            "width": 110
        },
        {
            "label": "Status", 
            "fieldname": "status", 
            "fieldtype": "Data", 
            "width": 80
        }
    ]

def get_data(filters):
    params = {
        "from_date": filters.get("from_date") if filters else None,
        "to_date": filters.get("to_date") if filters else None,
        "customer": filters.get("customer") if filters else None,
    }
    
    sql = """
        SELECT
            dn.name as do_no,
            DATE_FORMAT(dn.posting_date, '%%d-%%b-%%y') AS do_date,
            dn.customer_name,
            dn.currency,
            dn.status,
            dni.item_code,
            dni.item_name,
            dni.description,
            dni.qty,
            dni.uom,
            dni.rate,
            dni.amount,
            
            (
                SELECT GROUP_CONCAT(DISTINCT sales_person SEPARATOR ', ') 
                FROM `tabSales Team` 
                WHERE parent = dn.name
            ) as salesman,
            
            (
                SELECT parent 
                FROM `tabSales Invoice Item` 
                WHERE delivery_note = dn.name 
                LIMIT 1
            ) as inv_no,

            (
                SELECT posting_date 
                FROM `tabSales Invoice` 
                WHERE name = (
                    SELECT parent 
                    FROM `tabSales Invoice Item` 
                    WHERE delivery_note = dn.name 
                    LIMIT 1
                )
            ) as inv_date

        FROM
            `tabDelivery Note` dn
        INNER JOIN
            `tabDelivery Note Item` dni ON dn.name = dni.parent
        WHERE
            dn.docstatus = 1
            AND (
                %(from_date)s IS NULL OR %(to_date)s IS NULL
                OR dn.posting_date BETWEEN %(from_date)s AND %(to_date)s
            )
            AND (%(customer)s IS NULL OR dn.customer = %(customer)s)
        ORDER BY
            dn.name ASC
    """

    raw_data = frappe.db.sql(sql, params, as_dict=True)
    
    final_data = []
    
    # Inisialisasi dictionary untuk menyimpan total per mata uang
    totals_by_currency = {}

    # Font Size 10px & line-height 1.1
    style_base = "font-size: 10; line-height: 1.1; color: #36414c;"
    style_bold = "font-size: 10; font-weight: bold; line-height: 1.1; color: #36414c;"

    
    for row in raw_data:
        # --- 1. HITUNG TOTAL PER CURRENCY ---
        currency = row.currency
        qty = flt(row.qty)
        amount = flt(row.amount)

        if currency not in totals_by_currency:
            totals_by_currency[currency] = {"total_qty": 0, "total_amount": 0}
        
        totals_by_currency[currency]["total_qty"] += qty
        totals_by_currency[currency]["total_amount"] += amount

        # --- 2. FORMAT DO INFO ---
        do_link = get_url_to_form("Delivery Note", row.do_no)
        formatted_do_date = formatdate(row.do_date) if row.do_date else "-"
        salesman_txt = row.salesman if row.salesman else "-"
        
        row["do_info"] = f"""
            <div style="{style_base}">
                <a href="{do_link}" style="{style_bold}">{row.do_no}</a><br>
                <span>{formatted_do_date}</span><br>
                <span>{salesman_txt}</span>
            </div>
        """

        # --- 3. FORMAT INVOICE INFO ---
        if row.inv_no:
            inv_link = get_url_to_form("Sales Invoice", row.inv_no)
            formatted_inv_date = formatdate(row.inv_date) if row.inv_date else "-"
            row["inv_info"] = f"""
                <div style="{style_base}">
                    <a href="{inv_link}" style="{style_bold}">{row.inv_no}</a><br>
                    <span>{formatted_inv_date}</span>
                </div>
            """
        else:
            row["inv_info"] = "-"

        # --- 4. FORMAT ITEM ---
        desc_text = row.description if row.description else row.item_name
        
        # Hapus Enter/Spasi kosong
        if desc_text:
            desc_text = " ".join(desc_text.split())
        
        if row.item_code == desc_text:
              item_html = f"<div style='{style_bold}'>{row.item_code}</div>"
        else:
            item_html = f"""
                <div style="{style_base}">
                    <span style="{style_bold}">{row.item_code}</span><br>
                    <span>{desc_text}</span>
                </div>
             """
        row["item_display"] = item_html

        # --- 5. FORMAT QTY & UOM ---
        row["qty_uom"] = f"""
            <div style="{style_base}">
                <span style="{style_bold}">{flt(row.qty)}</span><br>
                <span>{row.uom}</span>
            </div>
        """

        final_data.append(row)

    # --- 6. TAMBAHKAN BARIS TOTAL PER CURRENCY ---
    # Styling untuk baris total
    total_style = "font-size: 10; font-weight: bold; line-height: 1.2;"

    for currency, totals in totals_by_currency.items():
        total_row = {
            "do_info": f"<div style='{total_style}'>Total ({currency})</div>",
            "inv_info": "",
            "customer_name": "",
            "item_display": "",
            "currency": currency,
            "qty_uom": f"""
                <div style="{total_style}">
                    <span>{flt(totals['total_qty'])}</span>
                </div>
            """,
            "rate": "",
            "amount": totals['total_amount'],
            "status": "",
            "is_total_row": 1
        }

        final_data.append(total_row)

    return final_data
